### PR TITLE
[WIP] v5: Carousel improvements for easier customization

### PR DESCRIPTION
**Pull request by mdo**
**WIP, do not merge.**

This PR starts some work to revamp the carousel to use `currentColor`, allowing us to better use utilities to change text, control, and indicator colors. Can't recall where I left this off previously, so need to dig back in before I get a checklist of todos going. At the very least I'll need to consider some migration docs.

### CODE REVIEW

**Comment by mdo** [comment]
Cleaned some of this up for the new example I added many months ago. Any additional ideas for what to improve here for v5 @twbs/css-review?

File: scss/_carousel.scss [hunk 1]
```diff
@@ -158,21 +161,35 @@
...
+    &::after {
+      position: absolute;
+      top: -$carousel-indicator-hit-area-height;
+      right: -$carousel-indicator-hit-area-height / 2;
+      bottom: -$carousel-indicator-hit-area-height;
+      left: -$carousel-indicator-hit-area-height / 2;
+      display: block;
+      height: $carousel-indicator-hit-area-height * 3;
+      content: "";
+      background-color: rgba(255, 0, 0, .1);
```
Comment (MartijnCuppens) [inline scss/_carousel.scss]: Why red? 😃

File: scss/_carousel.scss [hunk 2]
```diff
@@ -123,18 +124,20 @@
...
   display: inline-block;
   width: $carousel-control-icon-width;
   height: $carousel-control-icon-width;
-  background: no-repeat 50% / 100% 100%;
+  // fill: $carousel-control-icon-color;
+  // background: no-repeat 50% / 100% 100%;
 }
 .carousel-control-prev-icon {
-  background-image: escape-svg($carousel-control-prev-icon-bg);
+  // background-image: escape-svg($carousel-control-prev-icon-bg);
```
Comment (MartijnCuppens) [inline scss/_carousel.scss]: The controls are gone now?

File: scss/_carousel.scss [hunk 3]
```diff
@@ -158,21 +161,35 @@
   list-style: none;
 
   li {
-    box-sizing: content-box;
+    position: relative;
+    // box-sizing: content-box;
```
Comment (MartijnCuppens) [inline scss/_carousel.scss]: Could you remove these comments?

File: scss/_carousel.scss (hunk 1 above)
Comment (MartijnCuppens) [reply scss/_carousel.scss]: I think this shadow needs to be applied to the element itself

File: scss/_carousel.scss [hunk 4]
```diff
@@ -158,21 +161,35 @@
...
     @include transition($carousel-indicator-transition);
+
+    &::after {
+      position: absolute;
+      top: -$carousel-indicator-hit-area-height;
+      right: -$carousel-indicator-hit-area-height / 2;
+      bottom: -$carousel-indicator-hit-area-height;
+      left: -$carousel-indicator-hit-area-height / 2;
+      display: block;
+      height: $carousel-indicator-hit-area-height * 3;
```
Comment (MartijnCuppens) [inline scss/_carousel.scss]: Height is redundant since we already set top and bottom

File: scss/_variables.scss [hunk 5]
```diff
@@ -1066,31 +1066,33 @@ $breadcrumb-border-radius:          $border-radius !default;
...
+$carousel-indicator-width:           2rem !default;
+$carousel-indicator-height:          .25rem !default;
+$carousel-indicator-hit-area-height: .25rem !default;
+$carousel-indicator-spacer:          .25rem !default;
+$carousel-indicator-active-bg:       inherit !default;
 $carousel-indicator-transition:      opacity .6s ease !default;
 
 $carousel-caption-width:             70% !default;
-$carousel-caption-color:             $white !default;
+$carousel-caption-color:             inherit !default;
```
Comment (MartijnCuppens) [inline scss/_variables.scss]: Can we use `null` here?

File: scss/_variables.scss [hunk 6]
```diff
@@ -1066,31 +1066,33 @@ $breadcrumb-border-radius:          $border-radius !default;
...
 
-$carousel-control-icon-width:        20px !default;
+$carousel-control-icon-width:        3rem !default;
+$carousel-control-icon-color:        currentColor !default;
```
Comment (MartijnCuppens) [inline scss/_variables.scss]: This won't work. Background images don't know what the `currentColor` is. Alternatives:
- Use an `<svg>` in the DOM
- Work with a rotated pseudo element which is rotated

File: scss/_carousel.scss [hunk 7]
```diff
@@ -158,21 +161,35 @@
...
-    background-clip: padding-box;
+    // background-color: $carousel-indicator-active-bg;
+    // background-clip: padding-box;
+    border-bottom: $carousel-indicator-height solid;
     // Use transparent borders to increase the hit area by 10px on top and bottom.
-    border-top: $carousel-indicator-hit-area-height solid transparent;
-    border-bottom: $carousel-indicator-hit-area-height solid transparent;
+    // border-top: $carousel-indicator-hit-area-height solid transparent;
+    // border-bottom: $carousel-indicator-hit-area-height solid transparent;
     opacity: .5;
```
Comment (MartijnCuppens) [inline scss/_carousel.scss]: Probably interesting to make this a variable

File: scss/_carousel.scss [hunk 8]
```diff
@@ -158,21 +161,35 @@
...
   li {
-    box-sizing: content-box;
+    position: relative;
+    // box-sizing: content-box;
     flex: 0 1 auto;
     width: $carousel-indicator-width;
     height: $carousel-indicator-height;
     margin-right: $carousel-indicator-spacer;
     margin-left: $carousel-indicator-spacer;
     text-indent: -999px;
```
Comment (MartijnCuppens) [inline scss/_carousel.scss]: Do we really need this? No text is added in the thumbnails anyway. If it needs text, we can work with `aria-labels`?

File: scss/_variables.scss [hunk 9]
```diff
@@ -1066,31 +1066,33 @@ $breadcrumb-border-radius:          $border-radius !default;
 
 // Carousel
 
-$carousel-control-color:             $white !default;
+$carousel-color:                     $white !default;
+
+$carousel-control-color:             inherit !default;
```
Comment (MartijnCuppens) [inline scss/_variables.scss]: Can we use `null` here?

File: scss/_carousel.scss [hunk 10]
```diff
@@ -13,6 +13,7 @@
 
 .carousel {
   position: relative;
+  color: $carousel-color;
```
Comment (MartijnCuppens) [inline scss/_carousel.scss]: I like this. Maybe we can also add an example with light slides with a `.text-dark` added to the carousel?

File: site/content/docs/4.3/components/carousel.md [hunk 11]
```diff
@@ -208,6 +208,54 @@ Add `data-interval=""` to a `.carousel-item` to change the amount of time to del
...
+    </div>
+    <div class="carousel-item">
+      {{< placeholder width="800" height="400" class="bd-placeholder-img-lg d-block w-100" color="#333" background="#555" text="Third slide" >}}
+      <div class="carousel-caption d-none d-md-block">
+        <h5>Third slide label</h5>
+        <p>Praesent commodo cursus magna, vel scelerisque nisl consectetur.</p>
+      </div>
+    </div>
+  </div>
+  <a class="carousel-control-prev" href="#carouselExampleColors" role="button" data-slide="prev">
```
Comment (MartijnCuppens) [inline site/content/docs/4.3/components/carousel.md]: `<a role="button">`, looks nasty. Can we just use `<button>`s here?

File: scss/_carousel.scss (hunk 7 above)
Comment (MartijnCuppens) [reply scss/_carousel.scss]: Would close https://github.com/twbs/bootstrap/issues/29332

**Review by samgomena — APPROVED** [review]

File: site/content/docs/4.3/components/carousel.md (hunk 11 above)
Comment (mdo) [reply site/content/docs/4.3/components/carousel.md]: This is how the carousel works today, so this isn't any kind of regression. I'm down to use buttons here, but I'd need help to rewrite the JS that attaches these anchors to the appropriate carousel.

**Comment by mdo** [comment]
Closing as stale.

**Pull request closed without merging**